File: apps/umi-pos/ios/Runner/DeviceKeySigner.swift
import Flutter
import Foundation
import Security

/// iOS Secure Enclave backing for the POS device key. The private key is
/// generated inside the Secure Enclave and never leaves it; signing happens in
/// the enclave. It answers the `co.umiconsulting.umi_pos/device_key`
/// MethodChannel that `MethodChannelKeystore` (Dart) calls.
///
/// Contract, matching the Dart side:
///  - `ensurePublicKey` -> the public key as X.509 SubjectPublicKeyInfo (SPKI)
///    DER. `SecKeyCopyExternalRepresentation` gives the raw X9.63 point
///    (0x04‖X‖Y), so we prepend the fixed P-256 SPKI header to it.
///  - `sign(message)` -> an ECDSA SHA-256 signature (ASN.1 DER). The Dart side
///    normalizes DER to the 64-byte raw r‖s the server verifies.
///
/// REVIEW STATUS: written to the documented Secure Enclave contract but NOT
/// compiled or run in the build environment (no macOS/Xcode/device there). It
/// must be built and exercised on a device before it is trusted.
final class DeviceKeySigner {
  static let channelName = "co.umiconsulting.umi_pos/device_key"
  private let keyTag = "co.umiconsulting.umi_pos.device_key".data(using: .utf8)!

  // The 26-byte ASN.1 SubjectPublicKeyInfo header for an EC P-256 (prime256v1)
  // public key. The Secure Enclave hands back only the raw point; the server
  // reads SPKI, so we splice the two together.
  private let spkiHeader: [UInt8] = [
    0x30, 0x59, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86,
    0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08, 0x2a,
    0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07, 0x03,
    0x42, 0x00,
  ]

  static func register(with messenger: FlutterBinaryMessenger) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    let signer = DeviceKeySigner()
    channel.setMethodCallHandler { call, result in
      do {
        switch call.method {
        case "ensurePublicKey":
          result(FlutterStandardTypedData(bytes: try signer.ensurePublicKey()))
        case "sign":
          guard
            let args = call.arguments as? [String: Any],
            let message = (args["message"] as? FlutterStandardTypedData)?.data
          else { throw SignerError.badArguments }
          result(FlutterStandardTypedData(bytes: try signer.sign(message)))
        default:
          result(FlutterMethodNotImplemented)
        }
      } catch {
        result(FlutterError(code: "device_key_error",
                            message: "\(error)", details: nil))
      }
    }
  }

  private func loadPrivateKey() -> SecKey? {
    let query: [String: Any] = [
      kSecClass as String: kSecClassKey,
      kSecAttrApplicationTag as String: keyTag,
      kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
      kSecReturnRef as String: true,
    ]
    var item: CFTypeRef?
    guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess else {
      return nil
    }
    return (item as! SecKey)
  }

  private func ensurePrivateKey() throws -> SecKey {
    if let existing = loadPrivateKey() { return existing }

    guard let access = SecAccessControlCreateWithFlags(
      kCFAllocatorDefault,
      kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
      .privateKeyUsage, nil,
    ) else { throw SignerError.accessControl }

    let attributes: [String: Any] = [
      kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
      kSecAttrKeySizeInBits as String: 256,
      kSecAttrTokenID as String: kSecAttrTokenIDSecureEnclave,
      kSecPrivateKeyAttrs as String: [
        kSecAttrIsPermanent as String: true,
        kSecAttrApplicationTag as String: keyTag,
        kSecAttrAccessControl as String: access,
      ],
    ]
    var error: Unmanaged<CFError>?
    guard let key = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
      throw SignerError.keyCreation(error?.takeRetainedValue())
    }
    return key
  }

  private func ensurePublicKey() throws -> Data {
    let privateKey = try ensurePrivateKey()
    guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
      throw SignerError.noPublicKey
    }
    var error: Unmanaged<CFError>?
    guard let point = SecKeyCopyExternalRepresentation(publicKey, &error) as Data? else {
      throw SignerError.export(error?.takeRetainedValue())
    }
    // point is 0x04‖X‖Y (65 bytes); prepend the SPKI header to make it DER.
    return Data(spkiHeader) + point
  }

  private func sign(_ message: Data) throws -> Data {
    let privateKey = try ensurePrivateKey()
    var error: Unmanaged<CFError>?
    guard let signature = SecKeyCreateSignature(
      privateKey,
      .ecdsaSignatureMessageX962SHA256, // hashes the message, returns DER
      message as CFData, &error,
    ) as Data? else {
      throw SignerError.signing(error?.takeRetainedValue())
    }
    return signature
  }

  enum SignerError: Error {
    case badArguments
    case accessControl
    case keyCreation(CFError?)
    case noPublicKey
    case export(CFError?)
    case signing(CFError?)
  }
}

File: apps/umi-pos/android/app/src/main/kotlin/co/umiconsulting/umi_pos/DeviceKeySigner.kt
package co.umiconsulting.umi_pos

import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import io.flutter.plugin.common.MethodChannel
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.PrivateKey
import java.security.Signature
import java.security.spec.ECGenParameterSpec

/**
 * Android Keystore backing for the POS device key. The private key is generated
 * inside the Keystore (hardware-backed TEE, or StrongBox when the device has
 * one) and never leaves it; signing happens in the secure element. It answers
 * the `co.umiconsulting.umi_pos/device_key` MethodChannel that
 * `MethodChannelKeystore` (Dart) calls.
 *
 * Contract, matching the Dart side:
 *  - `ensurePublicKey` -> the public key as X.509 SubjectPublicKeyInfo (SPKI)
 *    DER bytes. `PublicKey.getEncoded()` returns exactly that.
 *  - `sign(message)` -> a SHA256withECDSA signature (ASN.1 DER). The Dart side
 *    normalizes DER to the 64-byte raw r‖s the server verifies.
 *
 * REVIEW STATUS: written to the documented AndroidKeyStore contract but NOT
 * compiled or run in the build environment (no Android SDK/device there). It
 * must be built and exercised on a device before it is trusted.
 */
class DeviceKeySigner(channel: MethodChannel) {
  init {
    channel.setMethodCallHandler { call, result ->
      try {
        when (call.method) {
          "ensurePublicKey" -> result.success(ensurePublicKey())
          "sign" -> {
            val message = call.argument<ByteArray>("message")
              ?: throw IllegalArgumentException("missing 'message'")
            result.success(sign(message))
          }
          else -> result.notImplemented()
        }
      } catch (e: Exception) {
        result.error("device_key_error", e.message, null)
      }
    }
  }

  private fun keyStore(): KeyStore =
    KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }

  private fun ensureKey(): PrivateKey {
    val store = keyStore()
    (store.getEntry(KEY_ALIAS, null) as? KeyStore.PrivateKeyEntry)?.let {
      return it.privateKey
    }
    val builder = KeyGenParameterSpec.Builder(KEY_ALIAS, KeyProperties.PURPOSE_SIGN)
      .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
      .setDigests(KeyProperties.DIGEST_SHA256)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
      // Prefer a discrete StrongBox secure element; fall back if absent.
      try {
        builder.setIsStrongBoxBacked(true)
      } catch (_: Exception) {
        builder.setIsStrongBoxBacked(false)
      }
    }
    val generator = KeyPairGenerator.getInstance(
      KeyProperties.KEY_ALGORITHM_EC, ANDROID_KEYSTORE,
    )
    return try {
      generator.initialize(builder.build())
      generator.generateKeyPair().private
    } catch (_: Exception) {
      // A device without StrongBox rejects the flag; retry TEE-backed.
      generator.initialize(
        KeyGenParameterSpec.Builder(KEY_ALIAS, KeyProperties.PURPOSE_SIGN)
          .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
          .setDigests(KeyProperties.DIGEST_SHA256)
          .build(),
      )
      generator.generateKeyPair().private
    }
  }

  private fun ensurePublicKey(): ByteArray {
    ensureKey()
    val cert = keyStore().getCertificate(KEY_ALIAS)
      ?: throw IllegalStateException("no certificate for device key")
    // X.509 SubjectPublicKeyInfo (SPKI) DER — what the Dart/server side expects.
    return cert.publicKey.encoded
  }

  private fun sign(message: ByteArray): ByteArray {
    val signature = Signature.getInstance("SHA256withECDSA")
    signature.initSign(ensureKey())
    signature.update(message)
    return signature.sign() // ASN.1 DER
  }

  companion object {
    const val CHANNEL = "co.umiconsulting.umi_pos/device_key"
    private const val ANDROID_KEYSTORE = "AndroidKeyStore"
    private const val KEY_ALIAS = "umi_pos_device_key"
  }
}
